vembrane/backend/backend_pysam.py

```diff
@@ -0,0 +1,273 @@
+        self._data = []
+        self._data_category = defaultdict(OrderedDict)
+        self._data_generic = dict()
+        for r in self._header.records:
+            if r.type == "GENERIC":
+                self._data_generic[r.key] = r.value
+                continue
+            d = dict(r)
+            self._data.append(d)
+            if "ID" in d:
+                self._data_category[r.type][d["ID"]] = d
+
+        # override numbers
+        for categorie, items in overwrite_number.items():
+            for key, value in items.items():
+                self._data_category[categorie][key]["Number"] = value
+
+    def contains_generic(self, key: str):
+        return key in self._data_generic
+
+    @property
+    def infos(self):
+        return self._data_category["INFO"]
+
+    @property
+    def formats(self):
+        return self._data_category["FORMAT"]
+
+    def __iter__(self):
+        self._iter_index = 0
+        return self
+
+    def __next__(self):
+        if self._iter_index >= len(self._data):
+            raise StopIteration
+        ret = self._data[self._iter_index]
+        self._iter_index += 1
+        return ret
+
+    @property
+    def samples(self):
+        return self._header.samples
+
+    @property
+    def filters(self):
+        return self._header.filters
+
+    @property
+    def records(self):
+        return self._data_category
+
+    # def add_meta(
+    #     self,
+    #     key: str,
+    #     value: Optional[str] = None,
+    #     items: Optional[List[Tuple[str, str]]] = None,
+    # ):
+    #     self._file.header.add_meta(key, value, items)
+
+    def add_generic(self, key: str, value: str):
+        self._header.add_meta(key, value)
+
+    def add_filter(self, id: str, description: str):
+        self._header.add_meta(
+            key="FILTER", items=[("ID", id), ("Description", description)]
+        )
+
+
+class PysamWriter(VCFWriter):
+    def __init__(self, filename: str, fmt: str, template: VCFReader):
+        self.filename = filename
+        self._file = pysam.VariantFile(
+            self.filename, f"w{fmt}", header=template.header._header
+        )
+        self._header = PysamHeader(self)
+
+    def write(self, record: PysamRecord):
+        self._file.write(record._record)
+
+
+def get_end(record: VariantRecord):
+    if is_bnd_record(record):
+        return NA
+    else:
+        # record.stop is pysams unified approach to get the end position of a variant.
+        # It either considers the alt allele or the END field, depending on the record.
+        # Stop is 0-based, but for consistency with POS we convert into 1-based.
+        # Since for 1-based coordinates, the expectation is that END is inclusive
+        # instead of exclusive (as it is with 0-based), we do not need to add 1.
+        return record.stop
+
+
+def is_bnd_record(record: VariantRecord) -> bool:
+    return "SVTYPE" in record.info and record.info.get("SVTYPE", None) == "BND"

```

Comment:
This is a code duplication (see method of same name in `common.py`) and can probably be removed.

Reply:
Yes it it. I am avoiding a circular import here. The only solution I can come up with is splitting the common file into multiple files. Any ideas are welcome!

Reply:
Oh, right. Hm, I guess we can just leave it as is for now

Reply:
I just move the functions to VCFRecord